lib/zzhc/zzhc.c

```diff
@@ -0,0 +1,600 @@
+		if (rc < 0) {
+			LOG_DBG("Error reading from socket, errno=%d", errno);
+			ctx->state = STATE_RETRY_TIMEOUT;
+			break;
+		}
+
+		if (rc == 0) {
+			LOG_DBG("Peer closed connection!");
+			ctx->state = STATE_HTTP_RESP_RECVD;
+			break;
+		}
+		break;
+
+	case STATE_HTTP_RESP_RECVD:
+		LOG_DBG("STATE_HTTP_RESP_RECVD");
+
+		/* Parse HTTP response. */
+		if (check_http_payload(ctx)) {
+			/* Try to write ICCID to local storage. */
+			rc = save_iccid(ctx->ueiccid, ICCID_LEN);
+			if (rc <= 0) {
+				LOG_DBG("save_iccid() = %d", rc);
+			}
+
+			ctx->state = STATE_END;
+		} else {
+			ctx->state = STATE_RETRY_TIMEOUT;
+		}
+		break;
+
+	case STATE_RETRY_TIMEOUT:
+		LOG_DBG("STATE_RETRY_TIMEOUT");
+
+		/* Cleanup */
+		disconnect(ctx);
+		ctx->http_pkt_len = 0;
+		memset(ctx->http_pkt, 0, HTTP_PKT_LEN);
+
+		if (ctx->retry_cnt >= RETRY_CNT_MAX) {
+			ctx->state = STATE_END;
+			break;
+		}
+
+		sleep(RETRY_TIMEOUT);
+		ctx->retry_cnt++;
+		ctx->state = (ctx->registered)
+			? STATE_START : STATE_WAIT_FOR_REG;
+		break;
+
+	case STATE_END:
+		LOG_DBG("STATE_END");
+
+		/* Clean-up */
+		disconnect(ctx);
+		deregister_handler(ctx, at_notif_handler);
+		ext_uninit(ctx);
+		this_free(ctx);
+		return -ECANCELED;
+
+	default:
+		LOG_DBG("Unknown state=%d", (int)ctx->state);
+		return -ECANCELED;
+	}
+	return 0;
+}
+
+/**@brief Thread function. */
+static void thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	struct zzhc *ctx = (struct zzhc *)arg1;
+
+	while (true) {
+		if (fsm(ctx) == -ECANCELED) {
+			return;
+		}
+	}
+}
+
+/**@brief Function for initializing self-registration daemon. */
+static int zzhc_init(struct device *arg)
+{
+	int rc;
+	struct zzhc *ctx;
+
+	/* Unconditionally subscribe +CEREG events */
+	if (at_cmd_xfer("AT+CEREG=5", NULL, 0) != 0) {
+		LOG_DBG("Can't subscribe to +CEREG events.");
+		return -EIO;
+	}
+
+	/* Allocate resource. */
+	ctx = (struct zzhc *)this_malloc(sizeof(struct zzhc));
+	if (ctx == NULL) {
+		LOG_DBG("Can't allocate memory for context.");
+		return -ENOBUFS;
+	}
+	arg->driver_data = ctx;
+
+	/* Initial values */
+	memset(ctx, 0, sizeof(struct zzhc));
+	ctx->state      = STATE_INIT;
+	ctx->registered = false;
+	ctx->sim_ready  = false;
+	ctx->sock_fd    = -1;
+	ctx->retry_cnt  = 0;
+
+	/* Initialize external modules */
+	rc = ext_init(ctx);
+	if (rc != 0) {
+		this_free(ctx);
+		return rc;
+	}
+
+	/* Try to read ICCID from local storage. */
+	rc = load_iccid(ctx->ueiccid, ICCID_LEN);
+	if (rc <= 0) {
+		LOG_DBG("load_iccid() = %d", rc);
+	}
+
+	/* Create thread */
+	/* Note: this function either succeed or result in kernel panic */
+	thread_create(ctx->thread, NULL, thread_fn, ctx);
+	return 0;
+}
+
+/**@brief Dummy API structure for DEVICE_AND_API_INIT
+ *
+ * @note  Otherwise, device_get_binding() will return NULL.
+ */
+struct api {
+	void *ptr;
+};
+
+static struct api const api = {NULL};
+
+/**@brief Device definition for zzhc */
+DEVICE_AND_API_INIT(zzhc, "ZZHC", zzhc_init, NULL, NULL, APPLICATION,
```

Comment:
Yes, it could be done.

The original design of at_cmd notification does not allow for passing of context pointers to receiving module. The intention of using DEVICE_AND_API_INIT() is to make sure the at-notification handler can get the pointer to context through the use of device_get_binding(). However, now the new design in at_notif allows for passing of pointer to context. Therefore, we do not need DEVICE_AND_API_INIT().

There is another consideration which I do not know if the K_THREAD_DEFINE() approach supports or not. It is that zzhc_init() has to happen after at_cmd, at_notif, nvs and settings are ready. DEVICE_AND_API_INIT() and SYS_INIT() allows for such. Can the K_THREAD_DEFINE()() also support so?

Reply:
Yeah, I see your point... As for the `K_THREAD_DEFINE()`, all threads are fired up just before the main() starts, so after all initializations are complete, so there should be no harm in initializing within a thread itself. 

But that's how it works, I haven't found any formal confirmation of this behavior in the docs. `SYS_INIT()` sounds fine for library initialization as well, I forgot about this function. So I leave it up to you whether to use `SYS_INIT()`, or initialize in the thread directly.

Reply:
I vote for K_THREAD_DEFINE(), because it is simpler than SYS_INIT().
We also need to care about the startup sequence of SYS_INIT() any more.

Reply:
Done in the last push on 4th Nov 2019.